src/shared/fork-journal.c

```diff
@@ -0,0 +1,203 @@
+#include "parse-util.h"
+#include "process-util.h"
+#include "signal-util.h"
+#include "socket-util.h"
+#include "strv.h"
+
+static int on_child_exit(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        PidRef *child = ASSERT_PTR(userdata);
+
+        assert(si->si_pid == child->pid);
+
+        log_debug("Child " PID_FMT " died before sending notification message.", child->pid);
+        return -EPROTO;
+}
+
+static int on_child_notify(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        PidRef *child = ASSERT_PTR(userdata);
+        int r;
+
+        _cleanup_strv_free_ char **msg = NULL;
+        _cleanup_(pidref_done) PidRef sender = PIDREF_NULL;
+        r = notify_recv_strv(fd, &msg, /* ret_ucred= */ NULL, &sender);
+        if (r == -EAGAIN)
+                return 0;
+        if (r < 0)
+                return r;
+
+        if (!pidref_equal(child, &sender)) {
+                log_warning("Received notification message from unexpected process " PID_FMT " (expected " PID_FMT "), ignoring.",
+                            sender.pid, child->pid);
+                return 0;
+        }
+
+        if (strv_find(msg, "READY=1"))
+                return sd_event_exit(sd_event_source_get_event(s), EXIT_SUCCESS);
+
+        const char *e = strv_find_startswith(msg, "ERRNO=");
+        if (e) {
+                int error;
+
+                r = safe_atoi(e, &error);
+                if (r < 0) {
+                        log_debug_errno(r, "Received invalid ERRNO= notification message, ignoring: %m");
+                        return 0;
+                }
+                if (error <= 0) {
+                        log_debug("Received non-positive ERRNO= notification message, ignoring: %m");
+                        return 0;
+                }
+
+                return sd_event_exit(sd_event_source_get_event(s), -error);
+        }
+
+        return 0;
+}
+
+int journal_fork(RuntimeScope scope, Set **pids, const char *unit) {
+        int r;
+
+        assert(scope >= 0);
+        assert(scope < _RUNTIME_SCOPE_MAX);
+        assert(unit);
+
+        if (!is_main_thread())
+                return -EPERM;
+
+        _cleanup_close_ int notify_fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (notify_fd < 0)
+                return log_debug_errno(errno, "Failed to allocate AF_UNIX socket for notifications: %m");
+
+        r = setsockopt_int(notify_fd, SOL_SOCKET, SO_PASSCRED, true);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to enable SO_PASSCRED: %m");
+
+        r = setsockopt_int(notify_fd, SOL_SOCKET, SO_PASSPIDFD, true);
+        if (r < 0)
+                log_debug_errno(r, "Failed to enable SO_PASSPIDFD, ignoring: %m");
+
+        /* Pick an address via auto-bind */
+        union sockaddr_union sa = {
+                .sa.sa_family = AF_UNIX,
+        };
+        if (bind(notify_fd, &sa.sa, offsetof(union sockaddr_union, un.sun_path)) < 0)
```

Comment:
SOCKADDR_UN_LEN() ?

Reply:
no, that wouldn't work. we want the kernel's autobind logic to trigger and we can only do that if we pass a zero sized path. But SOCKADDR_UN_LEN() would pass a 1 sized one, because it would interpret this as an abstract namespace sockaddr.

Or in other words: a sockaddr_un set up for an abstract namespace sockaddr with a zero length address has the same binary representation as the special zero-length sockaddr to trigger auto-bind. We cannot dtermine the length from the strucure itself hence, and must specify it explicitly.